kernel/src/utilities/capability_ptr.rs

```diff
@@ -6,20 +6,17 @@
 
 use core::fmt::{Formatter, LowerHex, UpperHex};
 use core::ops::AddAssign;
+use core::ptr::null;
 
 use super::machine_register::MachineRegister;
 
 /// A pointer to userspace memory with implied authority.
 ///
-/// A [`CapabilityPtr`] points to memory a userspace process may be
-/// permitted to read, write, or execute. It is sized exactly to a
-/// CPU register that can pass values between userspace and the kernel.
-/// Because it is register sized, [`CapabilityPtr`] is guaranteed to be
-/// at least the size of a word ([usize]) [^note1]. Operations on the
-/// pointer may affect permissions, e.g. offsetting the pointer beyond
-/// the bounds of the memory object invalidates it. Like a `*const
-/// ()`, a [`CapabilityPtr`] may also "hide" information by storing a
-/// word of data with no memory access permissions.
+/// A [`CapabilityPtr`] points to memory a userspace process may be permitted to
+/// read, write, or execute. It is sized exactly to a CPU register that can pass
+/// values between userspace and the kernel [^note1]. Operations on the pointer
+/// may affect permissions, e.g. offsetting the pointer beyond the bounds of the
+/// memory object invalidates it.
```

Comment:
```suggestion
/// memory object may invalidate it.
```

I think we should leave this as "possible", since on many systems CapabilityPtr won't actually support this behavior.

Reply:
This is tricky, because regardless of `CapabilityPtr` this operation may _also_ invalidate regular pointers in Rust (where subsequent dereferences could be unsound):

> If the computed offset is non-zero, then self must be [derived from](https://doc.rust-lang.org/stable/std/ptr/index.html#provenance) a pointer to some [allocated object](https://doc.rust-lang.org/stable/std/ptr/index.html#allocated-object), and the entire memory range between self and the result must be in bounds of that allocated object. In particular, this range must not “wrap around” the edge of the address space.

From https://doc.rust-lang.org/stable/std/primitive.pointer.html#safety-3.

Either way, I think that the proposed wording is strictly better, because we shouldn't overpromise and saying "may" does not preclude "will".

Reply:
> I think we should leave this as "possible", since on many systems CapabilityPtr won't actually support this behavior.

When I made this PR, I thought that having overly-conservative function documentation was fine. That is, if the docs say an operation will invalidate the authority, but the implementation only sometimes invalidates it, that's fine. However, upon further thought I realized there's an issue there: since CapabilityPtr is used for security, code might rely on a "this will invalidate the authority" guarantee in a way that results in a vulnerability if the authority is only sometimes invalidated. Therefore yes, we need to leave this as "possible" for correctness.

I'll send a PR correcting this when I get the chance.

> This is tricky, because regardless of `CapabilityPtr` this operation may _also_ invalidate regular pointers in Rust (where subsequent dereferences could be unsound):
> 
> > If the computed offset is non-zero, then self must be [derived from](https://doc.rust-lang.org/stable/std/ptr/index.html#provenance) a pointer to some [allocated object](https://doc.rust-lang.org/stable/std/ptr/index.html#allocated-object), and the entire memory range between self and the result must be in bounds of that allocated object. In particular, this range must not “wrap around” the edge of the address space.
> 
> From https://doc.rust-lang.org/stable/std/primitive.pointer.html#safety-3.

Calling `offset` on a Rust raw pointer with invalid arguments results in immediate UB, not UB on later dereference. Here's a [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9342da595faada9f57f0e007ed8be723) that steps through some of the failure modes here -- use the Tools menu to run the code under Miri to see the UB messages.

Reply:
Fixed.